lib/privileges.ts

```diff
@@ -377,3 +378,416 @@ export const checkPrivilegesAsBoolean = (
     return false;
   }
 };
+
+const _getSubject = async (subject: { type: Extract<Subject, string>; id: string }) => {
+  if (subject.id === "all") {
+    return {};
+  }
+  switch (subject.type) {
+    case "User":
+      return prisma.user.findUniqueOrThrow({
+        where: {
+          id: subject.id,
+        },
+        select: {
+          id: true,
+          name: true,
+          email: true,
+          lastLogin: true,
+          active: true,
+        },
+      });
+    case "FormRecord":
+      return prisma.template.findUniqueOrThrow({
+        where: {
+          id: subject.id,
+        },
+        select: {
+          id: true,
+          created_at: true,
+          updated_at: true,
+          name: true,
+          isPublished: true,
+          users: {
+            select: {
+              id: true,
+              name: true,
+              email: true,
+              active: true,
+            },
+          },
+        },
+      });
+    case "Privilege":
+      return prisma.privilege.findUniqueOrThrow({
+        where: {
+          id: subject.id,
+        },
+      });
+    case "Setting":
+      return prisma.setting.findUniqueOrThrow({
+        where: {
+          internalId: subject.id,
+        },
+      });
+    case "Flag":
+      return { id: subject.id, value: await checkOne(subject.id) };
+    default:
+      throw new Error("Subject type not valid for privilege check");
+  }
+};
+
+const _retrieveSubjects = async (
+  cache: Map<string, unknown>,
+  subject: {
+    type: Extract<Subject, string>;
+    scope: "all" | { subjectId: string } | { subjectIds: string[] };
+  }
+) => {
+  if (subject.scope === "all") {
+    return [{}];
+  }
+  if ("subjectId" in subject.scope) {
+    const cachedItem = cache.get(`${subject.type}:${subject.scope.subjectId}`);
+    if (cachedItem) return [cachedItem];
+
+    const item = await _getSubject({ type: subject.type, id: subject.scope.subjectId });
```

Comment:
Do we need a try catch somewhere here or are we expecting Prisma to catch?

Reply:
Instead of catching and rethrowing in every function the thought was to let it bubble up to `authorizationCheck` where it's caught and then transformed into an `AccessControlError`.

Reply:
okay